tmail-backend/jmap/extensions/src/main/scala/com/linagora/tmail/james/jmap/method/PublicAssetGetMethod.scala

```diff
@@ -0,0 +1,91 @@
+import org.apache.james.jmap.core.Id.Id
+import org.apache.james.jmap.core.Invocation.{Arguments, MethodName}
+import org.apache.james.jmap.core.{AccountId, Id, Invocation, SessionTranslator, UuidState}
+import org.apache.james.jmap.json.ResponseSerializer
+import org.apache.james.jmap.method.{InvocationWithContext, MethodRequiringAccountId}
+import org.apache.james.jmap.routes.SessionSupplier
+import org.apache.james.mailbox.MailboxSession
+import org.apache.james.metrics.api.MetricFactory
+import org.reactivestreams.Publisher
+import play.api.libs.json.JsObject
+import reactor.core.scala.publisher.{SFlux, SMono}
+
+import scala.util.Try
+
+class PublicAssetGetMethod @Inject()(val publicAssetRepository: PublicAssetRepository,
+                                     val metricFactory: MetricFactory,
+                                     val sessionTranslator: SessionTranslator,
+                                     val sessionSupplier: SessionSupplier) extends MethodRequiringAccountId[PublicAssetGetRequest] {
+
+  override val methodName: Invocation.MethodName = MethodName("PublicAsset/get")
+
+  override val requiredCapabilities: Set[CapabilityIdentifier] = Set(JMAP_CORE, LINAGORA_PUBLIC_ASSETS)
+
+  override def getRequest(mailboxSession: MailboxSession, invocation: Invocation): Either[Exception, PublicAssetGetRequest] =
+    Serializer.deserializeGetRequest(invocation.arguments.value)
+      .asEither.left.map(ResponseSerializer.asException)
+
+  override def doProcess(capabilities: Set[CapabilityIdentifier],
+                         invocation: InvocationWithContext,
+                         mailboxSession: MailboxSession,
+                         request: PublicAssetGetRequest): Publisher[InvocationWithContext] =
+    getPublicAssetResponse(mailboxSession.getUser, request)
+      .map(response => Invocation(
+        methodName = methodName,
+        arguments = Arguments(Serializer.serialize(response).as[JsObject]),
+        methodCallId = invocation.invocation.methodCallId))
+      .map(InvocationWithContext(_, invocation.processingContext))
+
+  private def getPublicAssetResponse(username: Username, request: PublicAssetGetRequest): SMono[PublicAssetGetResponse] = {
+    request.ids match {
+      case None => getAllAssets(username, request.accountId)
+      case Some(ids) => getSpecificAssets(username, request.accountId, ids)
+    }
+  }
+
+  private def getAllAssets(username: Username, accountId: AccountId): SMono[PublicAssetGetResponse] =
+    SFlux.fromPublisher(publicAssetRepository.list(username))
+      .collectSeq()
+      .map(seq => PublicAssetGetResponse(accountId,
+        state = UuidState.INSTANCE,
+        seq,
+        Seq.empty))
+
+  private def getSpecificAssets(username: Username, accountId: AccountId, ids: Set[Id]): SMono[PublicAssetGetResponse] = {
+    val parsedIds: Set[Either[(Id, IllegalArgumentException), PublicAssetId]] =
+      ids.map(id => Try(PublicAssetId(UUID.fromString(id)))
```

Comment:
we can write `PublicAssetId.fromString(input) : Either[IllegalArgumentException, PublicAssetId]`
It can be re-use for another method

Reply:
updated